oo7-daemon/src/service/service.rs

```diff
@@ -0,0 +1,191 @@
+//  org.freedesOnceCellktop.Secret.Service
+
+use std::collections::HashMap;
+
+use oo7::{dbus::api::Properties, portal::Item};
+use serde::Serialize;
+use zbus::{dbus_interface, fdo, zvariant, Error, ObjectServer, SignalContext};
+use zvariant::{ObjectPath, OwnedObjectPath, OwnedValue, Type, Value};
+
+use crate::{
+    service::{collection::Collection, session::Session},
+    KEYRING,
+};
+
+#[derive(Serialize, Type, Debug)]
+pub struct Service {
+    pub collections: Vec<Collection>,
+}
+
+#[dbus_interface(name = "org.freedesktop.Secret.Service")]
+impl Service {
+    pub async fn open_session(
+        &self,
+        // algorithm: &str,
+        input: Value<'_>,
+    ) -> fdo::Result<(OwnedValue, OwnedObjectPath)> {
+        // WIP: not complete
+        let session = Session::new(ObjectPath::try_from("/s1").unwrap().into()).await;
+        Ok((
+            OwnedValue::try_from(input).unwrap(), // avoid using unwrap
+            ObjectPath::try_from(session.path().await).unwrap().into(),
+        ))
+    }
+
+    #[dbus_interface(out_args("collection", "prompt"))]
+    pub async fn create_collection(
+        &self,
+        properties: Properties,
+        alias: String,
+        #[zbus(object_server)] object_server: &ObjectServer,
+    ) -> fdo::Result<(ObjectPath, ObjectPath)> {
+        let _attributes = properties.attributes(); // expand and pass these to Collection:new
+        let collection = Collection::new("temp"); // temporarily
+        let path = format!("/org/freedesktop/secrets_/collection/{}", &alias);
+
+        let _ = object_server.at(path.clone(), collection).await;
+        let new_collection_path = ObjectPath::try_from(path).unwrap(); // temporarily
+        let prompt = ObjectPath::default().into(); // temp Prompt
+
+        Ok((new_collection_path, prompt))
+    }
+
+    #[dbus_interface(out_args("unlocked", "locked"))]
+    pub async fn search_items(
+        &self,
+        attributes: HashMap<&str, &str>,
+    ) -> fdo::Result<(Vec<Item>, Vec<Item>)> {
+        let items = match KEYRING.get().unwrap().search_items(&attributes).await {
+            Ok(i) => i,
+            Err(_) => todo!(),
+        };
+
+        let mut unlocked: Vec<Item> = Vec::new();
+        let mut locked: Vec<Item> = Vec::new();
+
+        for item in items {
+            let attributes = item.attributes();
+            if attributes.get("locked").is_some() {
+                // this if condition is probably wrong
+                // how to access &AttributeValue value
+                locked.push(item)
+            } else {
+                unlocked.push(item)
+            }
+        }
+
+        Ok((unlocked, locked))
+    }
+
+    #[dbus_interface(out_args("unlocked", "prompt"))]
+    pub async fn unlock(
+        &mut self,
+        objects: Vec<OwnedObjectPath>,
```

Comment:
You can use IsUnlockable here I guess. Also the function currently assumes the objects are collections which is not true. They can be items as well.

Reply:
With this change: https://github.com/bilelmoussaoui/oo7/blob/004f98a69f98456d86083b0b5f9264c4471c6c1e/daemon/src/service/service.rs#L212 
oo7-daemon is no longer responds to `busctl`. I usually run `busctl --user tree org.freedesktop.secrets_` after a major change. It's no longer there. I'm not sure what's causing this :)